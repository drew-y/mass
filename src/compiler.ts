import binaryen from "binaryen";
import { promises as fs } from "fs";
import { parse } from "./parser";
import { scanForEntities } from "./entity-scanner";
import { Assembler } from "./assembler";
import { analyseSemantics } from "./semantic-analyser";
import { Module } from "./module";
import { parse as pathParse } from "path";
import { walkDir } from "./helpers";

export async function compile(path: string): Promise<binaryen.Module> {
    const rootModule = new Module({ name: "root" });

    const stdModule = await buildModuleTree(`${__dirname}/../std`, rootModule);
    scanForEntities(stdModule);
    analyseSemantics(stdModule);

    const userModule = rootModule.sub("user");
    userModule.scope.import(stdModule.subModules["mod"].exports);
    await buildModuleTree(path, userModule);

    scanForEntities(userModule);
    analyseSemantics(userModule);

    const assembler = new Assembler();
    assembler.compile(rootModule);
    return assembler.mod;
}

async function buildModuleTree(path: string, parent: Module): Promise<Module> {
    const stats = await fs.lstat(path);
    const pathInfo = pathParse(path);
    const module = parent.sub(pathInfo.name);

    if (stats.isDirectory()) {
        const children = (await walkDir(path))
            .filter(item => item.type === "dir" || item.extension === ".dm");

        for (const child of children) {
            await buildModuleTree(child.path, module)
        }

        return module;
    }

    if (pathInfo.ext !== ".dm") {
        throw new Error(`File must have .dm extension. ${path}`);
    }

    const code = await fs.readFile(path, { encoding: "utf8" });
    module.ast = parse(code, module.scope);

    return module;
}
