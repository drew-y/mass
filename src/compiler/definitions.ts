
export interface VariableValue {
    kind: "var";
    identifier: string;
    type: number;
    mutable: boolean;
    isLocal?: boolean;
    localIndex?: number;
    flags: string[];
}

export interface MethodValue {
    kind: "method";
    identifier: string;
    /** Parameters representing a generic type. */
    typeParameters: number[];
    parameterTypes: number[];
    returnType: string;
    flags: string[];
}

export type Value = VariableValue | MethodValue;
