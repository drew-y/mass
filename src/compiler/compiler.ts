import binaryen from "binaryen";
import { ValueCollection } from "./values";
import { MethodValue, LocalValue } from "./definitions";
import {
    parse, Instruction, CallExpression, ReturnStatement, IfExpression, Assignment,
    MethodDeclaration, VariableDeclaration, WhileStatement, MatchExpression, BinaryExpression
} from "../parser";
import uniqid from "uniqid";

export function compile(code: string) {
    const ast = parse(code);
    const mod = new binaryen.Module();
    const ids = new ValueCollection();
    mod.autoDrop();
    mod.addFunctionImport("print", "imports", "print", binaryen.i32, binaryen.none);
    mod.addFunction("main", binaryen.none, binaryen.none, [], compileBlock({
        body: ast, mod, vals: ids, returnType: binaryen.none
    }));
    mod.addFunctionExport("main", "main");
    return mod;
}

function compileBlock({
    body, mod, vals, returnType, existingInstructions,
    additionalInstructions, context = "global"
}: {
    body: Instruction[],
    mod: binaryen.Module,
    vals: ValueCollection,
    returnType?: number,
    existingInstructions?: number[],
    additionalInstructions?: number[],
    context?: "global" | "method";
}): number {
    const block: number[] = existingInstructions ?? [];

    body.forEach(instruction => {
        if (instruction.kind === "variable-declaration") {
            compileVariableDeclaration({ instruction, vals, mod, block, context });
            return;
        }

        if (instruction.kind === "method-declaration") {
            compileMethodDeclaration(instruction, vals, mod);
            return;
        }

        if (instruction.kind === "assignment") {
            compileAssignment(instruction, mod, vals, block);
            return;
        }

        // TODO: make this an expression
        if (instruction.kind === "if-expression") {
            compileIfExpression(block, mod, instruction, vals);
            return;
        }

        if (instruction.kind === "while-statement") {
            compileWhileStatement(block, mod, instruction, vals);
            return;
        }

        if (instruction.kind === "match-expression") {
            compileMatchExpression(block, mod, instruction, vals);
            return;
        }

        if (instruction.kind === "return-statement") {
            compileReturn(block, mod, instruction, vals);
            return;
        }

        block.push(compileExpression(instruction, mod, vals));
    });

    if (additionalInstructions) block.push(...additionalInstructions);
    return mod.block("", block, returnType);
}

// TODO: Support non integer cases.
// TODO: Support patterns (ranges, strings, enum destructuring, etc.)
// TODO: Support default
// TODO: Document how this works. ASAP
function compileMatchExpression(block: number[], mod: binaryen.Module, instruction: MatchExpression, vals: ValueCollection) {
    const indexFunctionName = `match-${uniqid()}`;
    const cases: { name: string, case: number, expression: number }[] = [];
    for (const dCase of instruction.cases) {
        const name = JSON.stringify(dCase.case);
        cases.push({
            name,
            case: compileExpression(dCase.case, mod, vals),
            expression: compileExpression(dCase.expression, mod, vals)
        });
    }

    // Build the match indexing function

    const matchBlock: number[] = [
        mod.local.set(0, compileExpression(instruction.value, mod, vals))
    ];

    cases.forEach((cCase, index) => {
        // If the match value is equal to the case, return the block index of the case's expression.
        matchBlock.push(mod.if(
            mod.i32.eq(cCase.case, mod.local.get(0, binaryen.i32)),
            mod.return(mod.i32.const(index + 1))
        ))
    });

    matchBlock.push(mod.i32.const(0));

    mod.addFunction(indexFunctionName, binaryen.createType([]), binaryen.i32, [binaryen.i32], mod.block("", matchBlock, binaryen.i32));

    // Convert the 1D cases array to a hierarchical set of blocks, last one containing the switch (br_table).
    // TODO: Make this iterative.
    function makeBlockTree(caseIndex = 0): number {
        const cCase = cases[caseIndex];

        if (cCase) {
            return mod.block(cCase.name, [
                makeBlockTree(caseIndex + 1),
                cCase.expression,
                mod.br("match")
            ]);
        }

        return mod.block("matcher", [
            mod.switch(
                [...cases.map(c => c.name), "matcher"],
                cases[0].name,
                mod.call(indexFunctionName, [], binaryen.i32)
            )
        ]);
    }

    block.push(mod.block("match", [makeBlockTree()]));
}

function compileReturn(block: number[], mod: binaryen.Module, instruction: ReturnStatement, vals: ValueCollection) {
    block.push(mod.return(compileExpression(instruction.expression, mod, vals)));
}

function compileIfExpression(block: number[], mod: binaryen.Module, instruction: IfExpression, vals: ValueCollection) {
    block.push(
        mod.if(compileExpression(instruction.condition, mod, vals),
            compileBlock({ body: instruction.body, mod, vals: vals }))
    );
}

function compileWhileStatement(block: number[], mod: binaryen.Module, instruction: WhileStatement, vals: ValueCollection) {
    block.push(
        mod.block("while", [
            mod.loop("loop",
                compileBlock({
                    body: instruction.body, mod, vals: vals,
                    existingInstructions: [
                        mod.br("while", mod.i32.ne(
                            compileExpression(instruction.condition, mod, vals),
                            mod.i32.const(1)
                        ))
                    ],
                    additionalInstructions: [
                        mod.br("loop")
                    ]
                })
            )
        ])
    );
}

function compileAssignment(instruction: Assignment, mod: binaryen.Module, vals: ValueCollection, block: number[]) {
    const id = instruction.identifier;
    const expr = compileExpression(instruction.expression, mod, vals);
    const val = vals.retrieve(id);

    if (val.kind === "method" || !val.mutable) {
        throw new Error(`${id} cannot be reassigned`);
    }

    if (val.kind === "local") {
        block.push(mod.local.set(val.index, expr));
        return;
    }

    block.push(mod.global.set(id, expr));
}

function compileMethodDeclaration(instruction: MethodDeclaration, vals: ValueCollection, mod: binaryen.Module) {
    const id = instruction.identifier;
    const internalVals = vals.clone();
    const params = instruction.parameters.map((param, index) => {
        const type = getTypeFromString(param.type!.identifier);
        internalVals.register({
            kind: "local",
            mutable: false,
            type, flags: [],
            id: param.identifier,
            index
        });
        return type;
    });

    const returnType = instruction.returnType ?
        getTypeFromString(instruction.returnType.identifier) :
        binaryen.none;

    const method: MethodValue = {
        kind: "method",
        id,
        parameters: params,
        returnType,
        flags: instruction.flags ?? []
    };
    internalVals.register(method);
    vals.register(method);

    const methodBody = compileBlock({
        body: instruction.body,
        mod,
        vals: internalVals,
        returnType,
        context: "method",
    });
    const additionalLocals = internalVals.getNonParameterLocalTypes();

    mod.addFunction(id, binaryen.createType(params), returnType, additionalLocals, methodBody);
}

function compileVariableDeclaration({
    instruction, vals, mod, block, context = "global"
}: {
    instruction: VariableDeclaration, vals: ValueCollection, mod: binaryen.Module, block: number[],
    context?: "global" | "method"
}) {
    const id = instruction.identifiers[0];
    const type = instruction.type ?
        getTypeFromString(instruction.type.identifier) :
        inferType(instruction.initializer!, vals);

    if (context === "method") {
        vals.register({
            kind: "local",
            id, type,
            flags: instruction.flags,
            nonParameter: true,
            mutable: instruction.flags.includes("var"),
            index: 0
        });

        if (instruction.initializer) {
            const local = vals.retrieve(id) as LocalValue;
            block.push(mod.local.set(
                local.index,
                compileExpression(instruction.initializer, mod, vals)
            ))
        }

        return;
    }

    mod.addGlobal(id, type, true, globalInit(type, mod));

    vals.register({
        kind: "global",
        id,
        type,
        mutable: instruction.flags.includes("var"),
        flags: instruction.flags
    });

    if (instruction.initializer) {
        block.push(mod.global.set(id, compileExpression(instruction.initializer, mod, vals)));
    }
}

function compileExpression(expr: Instruction, mod: binaryen.Module, vals: ValueCollection): number {
    if (expr.kind === "int-literal") {
        return mod.i32.const(Number(expr.value));
    }

    if (expr.kind === "float-literal") {
        return mod.f32.const(Number(expr.value));
    }

    if (expr.kind === "bool-literal") {
        return mod.i32.const(expr.value ? 1 : 0);
    }

    if (expr.kind === "identifier") {
        const identifier = vals.retrieve(expr.value);

        if (identifier.kind === "local") {
            return mod.local.get(identifier.index, identifier.type);
        }

        if (identifier.kind === "global") {
            return mod.global.get(identifier.id, identifier.type);
        }

        throw new Error(`Unsupported identifier type in expression: ${identifier.kind}`);
    }

    if (expr.kind === "binary-expression") {
        return compileBinaryExpression(expr, mod, vals);
    }

    if (expr.kind === "call-expression") {
        // TODO: Add to vals as stdlib
        if (expr.identifier === "print") {
            return (mod.call as any)("print", [compileExpression(expr.arguments[0], mod, vals)], binaryen.none);
        }

        const val = vals.retrieve(expr.identifier);
        if (val.kind !== "method") throw new Error(`${expr.identifier} is not a method`);
        const args = expr.arguments.map(instr => compileExpression(instr, mod, vals));
        return (mod.call as any)(val.id, args, val.returnType);
    }

    throw new Error(`Invalid expression ${expr.kind}`);
}

function compileBinaryExpression(expr: BinaryExpression, mod: binaryen.Module, ids: ValueCollection): number {
    const arg1 = expr.arguments.shift()!;
    const arg2 = expr.arguments.shift()!;
    const type = inferType(arg1, ids); // Probably room for performance improvements here.
    const cArg1 = compileExpression(arg1, mod, ids);
    const cArg2 = compileExpression(arg2, mod, ids);
    const id = expr.identifier;

    if (type === binaryen.i32 && id === "+") {
        return mod.i32.add(cArg1, cArg2)
    }

    if (type === binaryen.f32 && id === "+") {
        return mod.f32.add(cArg1, cArg2)
    }

    if (type === binaryen.i32 && id === "-") {
        return mod.i32.sub(cArg1, cArg2)
    }

    if (type === binaryen.f32 && id === "-") {
        return mod.f32.sub(cArg1, cArg2)
    }

    if (type === binaryen.i32 && id === "*") {
        return mod.i32.mul(cArg1, cArg2)
    }

    if (type === binaryen.f32 && id === "*") {
        return mod.f32.mul(cArg1, cArg2)
    }

    if (type === binaryen.i32 && id === "/") {
        return mod.i32.div_s(cArg1, cArg2)
    }

    if (type === binaryen.f32 && id === "/") {
        return mod.f32.div(cArg1, cArg2)
    }

    if (type === binaryen.i32 && id === "<") {
        return mod.i32.lt_s(cArg1, cArg2)
    }

    if (type === binaryen.i32 && id === ">") {
        return mod.i32.gt_s(cArg1, cArg2)
    }

    if (type === binaryen.i32 && id === "<=") {
        return mod.i32.le_s(cArg1, cArg2)
    }

    if (type === binaryen.i32 && id === ">=") {
        return mod.i32.ge_s(cArg1, cArg2)
    }

    if (type === binaryen.f32 && id === "<") {
        return mod.f32.lt(cArg1, cArg2)
    }

    if (type === binaryen.f32 && id === ">") {
        return mod.f32.gt(cArg1, cArg2)
    }

    if (type === binaryen.f32 && id === "<=") {
        return mod.f32.le(cArg1, cArg2)
    }

    if (type === binaryen.f32 && id === ">=") {
        return mod.f32.ge(cArg1, cArg2)
    }

    throw new Error(`Unsupported add type: ${type}`);
}

function inferType(expression: Instruction, ids: ValueCollection): number {
    if (expression.kind === "call-expression") {
        if (["+", "-", "*", "/"].includes(expression.identifier)) {
            return inferType(expression.arguments[0], ids);
        }

        if (["==", ">", "<", ">=", "<=", "and", "ir"].includes(expression.identifier)) {
            return binaryen.i32;
        }

        const identifier = ids.retrieve(expression.identifier);
        if (identifier.kind !== "method" || !identifier.returnType) {
            throw new Error(`Unable to infer return type for ${expression.identifier}`);
        }
        return identifier.returnType;
    }

    if (expression.kind === "bool-literal") return binaryen.i32;
    if (expression.kind === "float-literal") return binaryen.f32;
    if (expression.kind === "int-literal") return binaryen.i32;
    if (expression.kind === "return-statement") return inferType(expression.expression, ids);
    if (expression.kind === "if-expression") {
        return inferType(expression.body[expression.body.length - 1], ids);
    }

    if (expression.kind === "identifier") {
        const id = ids.retrieve(expression.value);
        if (id.kind === "method") throw new Error("Unexpected identifier");
        return id.type;
    }


    throw new Error(`Unable to infer type for ${expression.kind}`);
}

function globalInit(type: number, mod: binaryen.Module): number {
    switch (type) {
        case binaryen.i32:
            return mod.i32.const(0);
        case binaryen.i64:
            return mod.i64.const(0, 0);
        case binaryen.f32:
            return mod.f32.const(0);
        case binaryen.f64:
            return mod.f64.const(0);
        default:
            throw new Error(`Unknown type: ${type}`);
    }
}

function getTypeFromString(typeIdentifier: string): number {
    if (typeIdentifier === "i32") return binaryen.i32;
    if (typeIdentifier === "f32") return binaryen.f32;
    if (typeIdentifier === "i64") return binaryen.i64;
    if (typeIdentifier === "f64") return binaryen.f64;
    throw new Error(`Unknown type: ${typeIdentifier}`);
}

function getMutability(flags: string[]): boolean {
    if (flags.includes("let")) return false;
    if (flags.includes("var")) return true;
    throw new Error("Passed flags do not include mutability modifiers");
}
