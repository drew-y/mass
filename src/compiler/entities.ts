import uniqid from "uniqid";
import { Entity, EntityWithoutID } from "./definitions";
import { EventEmitter2 } from "eventemitter2";

export class Entities {
    private readonly emitter = new EventEmitter2();
    private readonly entities: Map<string, Entity> = new Map();

    entries(): Iterator<[string, Entity]> {
        return this.entities.entries();
    }

    /** Returns the ID of the entity. */
    add(entity: EntityWithoutID): string {
        const id = uniqid();
        const fullEntity = { id, ...entity } as Entity;
        this.entities.set(id, fullEntity);
        this.emitter.emit("Update", fullEntity);
        return id;
    }

    get(id: string): Entity {
        const entity = this.entities.get(id);

        if (!entity) {
            throw new Error(`Entity not found: ${id}`);
        }

        return entity;
    }

    update(entity: Partial<Entity>) {
        if (!entity.id) throw new Error("Entity must have ID");
        const updated: Entity = { ...this.entities.get(entity.id), ...entity } as Entity;
        this.entities.set(entity.id, updated);
        this.emitter.emit("Update", entity);
    }

    /** Emitted when an entity is added or updated */
    on(event: "Update", cb: (e: Entity) => void): void;
    on(event: string, cb: (...args: any[]) => void) {
        this.emitter.on(event, cb);
    }

    off(event: string, cb: (...args: any[]) => void) {
        this.emitter.off(event, cb);
    }
}
