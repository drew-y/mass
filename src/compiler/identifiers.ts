import { Identifier } from "./definitions";

export class IdentifiersCollection {
    private readonly identifiers: { [key: string]: Identifier };

    constructor(ids?: { [key: string]: Identifier }) {
        this.identifiers = ids ? JSON.parse(JSON.stringify(ids)) : {};
    }

    register(id: Identifier) {
        this.identifiers[id.identifier] = id;
    }

    retrieve(id: string): Identifier {
        const ident = this.identifiers[id];
        if (!ident) {
            throw new Error(`Unknown identifier: ${id}`);
        }

        return ident;
    }

    clone() {
        return new IdentifiersCollection(this.identifiers);
    }

    dir() {
        console.dir(this.identifiers);
    }
}
