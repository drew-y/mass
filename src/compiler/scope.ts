import { Entities } from "./entities";
import { Entity, FunctionEntity } from "./definitions";

export class Scope {
    /** entities withing this scope */
    protected readonly entities: Set<string> = new Set();

    /** All of the entities the module exports */
    readonly exports: string[] = [];

    readonly parent?: Scope;

    /** The scope allows local variables */
    isFnScope: boolean = false;

    /**
     * Defined locals in this scope. Only allowed if isMainFnScope.
     * Note: Having a local does not imply it is accessible from this scope.
     * It may have been defined in a sub scope of the function.
     */
    locals: string[] = [];

    constructor(parent?: Scope) {
        this.parent = parent;
    }

    /** Iterator of all entity IDs accessible from the scope, TODO: TRAVERSE PARENT SCOPES AND INCLUDE THEIR VALUES TOO */
    values() {
        return this.entities.values();
    }

    /** Add the entity as accessible from this scope */
    add(id: string) {
        this.entities.add(id);
    }

    /**
     * Adds the entity to the current scope, and as a local to the main function scope.
     * If we are not inside a function, throw an error.
     *
     * @param addToScope - For internal use only
     */
    addLocal(id: string, addToScope = true) {
        if (addToScope) this.add(id);

        if (this.isFnScope) {
            this.locals.push(id);
            return;
        }

        if (this.parent) {
            this.parent.addLocal(id, false);
            return;
        }

        throw new Error("Variable defined in invalid scope");
    }

    localsCount(): number {
        if (this.isFnScope) return this.locals.length;
        if (this.parent) return this.parent.localsCount();
        throw new Error("Not in a function");
    }

    export(id: string) {
        this.exports.push(id);
    }

    hasEntity(id: string): boolean {
        if (this.entities.has(id)) return true;

        if (this.parent) return this.parent.hasEntity(id);

        return false;
    }

    newSubScope(): Scope {
        return new Scope(this);
    }
}
