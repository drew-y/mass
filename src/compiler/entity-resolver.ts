import { Scope } from "./scope";
import { Entities } from "./entities";
import { Entity } from "./definitions";
import uniqid from "uniqid";

export class EntityResolver {
    private readonly entities: Entities;
    private readonly unresolved: Unresolved = new Map();

    constructor(entities: Entities) {
        this.entities = entities;
    }

    listUnresolved(): string[] {
        const labels: string[] = [];
        for (const [label, unresolvedForLabel] of this.unresolved.entries()) {
            if (unresolvedForLabel.size > 0) labels.push(label);
        }
        return labels;
    }

    resolveTypeEntity(req: Omit<EntityReq, "id">): void {
        const fullReq = { ...req, id: uniqid() };
        const { scope, label, resolver } = fullReq;

        // First, check if the entity already exists.
        const existing = scope.accessibleEntitiesWithLabel(label, this.entities);

        for (const entity of existing) {
            if (entity.kind !== "type") continue;
            return resolver(entity);
        }

        // An existing entity did not exist. Watch for it.
        this.addUnresolved(fullReq);
        const watcher = (entity: Entity) => {
            if (entity.kind !== "type") return;
            if (!entity.scope.hasEntity(entity.id)) return;
            resolver(entity);
            this.entities.off("Update", watcher);
            this.removeUnresolved(fullReq);
        };
        this.entities.on("Update", watcher);
    }

    resolveIdentifier(req: Omit<EntityReq, "id">): void {
        const fullReq = { ...req, id: uniqid() };
        const { scope, label, resolver } = fullReq;

        // First, check if the entity already exists.
        const existing = scope.accessibleEntitiesWithLabel(label, this.entities);

        for (const entity of existing) {
            if (entity.kind !== "local") continue;
            return resolver(entity);
        }

        // An existing entity did not exist. Watch for it.
        this.addUnresolved(fullReq);
        const watcher = (entity: Entity) => {
            if (entity.kind !== "local") return;
            if (!entity.scope.hasEntity(entity.id)) return;
            resolver(entity);
            this.entities.off("Update", watcher);
            this.removeUnresolved(fullReq);
        };
        this.entities.on("Update", watcher);
    }

    private addUnresolved(req: EntityReq) {
        if (!this.unresolved.has(req.label)) this.unresolved.set(req.label, new Map());
        const unresolvedForLabel = this.unresolved.get(req.label)!;
        unresolvedForLabel.set(req.id, req);
    }

    private removeUnresolved(req: EntityReq) {
        const unresolvedForLabel = this.unresolved.get(req.label);
        if (unresolvedForLabel) unresolvedForLabel.delete(req.id);
    }
}

/** Key is the unresolved label */
type Unresolved = Map<string, UnresolvedForLabel>;

/** Key is ID of the request. */
type UnresolvedForLabel = Map<string, EntityReq>;

type EntityReq = {
    id: string,
    label: string,
    scope: Scope,
    resolver: (e: Entity) => void
}
