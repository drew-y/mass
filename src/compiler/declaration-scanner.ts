import { Entities } from "./entities";
import { Scope } from "./scope";
import { AST, FunctionDeclaration, TypeDeclaration, Instruction, Identifier, VariableDeclaration, PropertyAccessExpression, ImplDeclaration, CallExpression, BinaryExpression, StructDeclaration, EnumDeclaration } from "../parser";
import { EntityResolver } from "./entity-resolver";

/**
 * Performs initial namespace and scoping resolution. Does not resolve function calls.
 *
 * This evolves:
 * 1. Creating entities from declarations.
 * 2. Adding entity ID's to the declaration instructions
 * 3. Resolving references to entities. I.E. adding ID's to identifiers
 * 4. Generating errors for unknown identifiers
 */
export class SemanticAnalyzer {
    private readonly entities: Entities;
    private readonly resolver: EntityResolver;

    constructor(entities: Entities) {
        this.entities = entities;
        this.resolver = new EntityResolver(entities);
    }

    analyze(ast: AST, scope: Scope) {
        for (const instruction of ast) {
            this.scanInstruction(instruction, scope);
        }
    }

    listUnresolved(): string[] {
        return this.resolver.listUnresolved();
    }

    private scanInstruction(instruction: Instruction, scope: Scope) {
        if (instruction.kind === "function-declaration") {
            this.scanFn(instruction, scope);
            return;
        }

        if (instruction.kind === "type-declaration") {
            this.scanType(instruction, scope);
            return;
        }

        if (instruction.kind === "enum-declaration") {
            this.scanType(instruction, scope);
            return;
        }

        if (instruction.kind === "struct-declaration") {
            this.scanType(instruction, scope);
            return;
        }

        if (instruction.kind === "identifier") {
            this.scanIdentifier(instruction, scope);
            return;
        }

        if (instruction.kind === "variable-declaration") {
            this.scanVariableDeclaration(instruction, scope);
            return;
        }

        if (instruction.kind === "property-access-expression") {
            this.scanPropertyAccessExpression(instruction, scope);
            return;
        }

        if (instruction.kind === "impl-declaration") {
            this.scanImplDeclaration(instruction, scope);
            return;
        }

        if (instruction.kind === "call-expression" || instruction.kind === "binary-expression") {
            this.scanCallExpression(instruction, scope);
            return;
        }

        if (instruction.kind === "block-expression") {
            instruction.body.forEach(val => this.scanInstruction(val, scope));
            return;
        }

        if (instruction.kind === "enum-variant") {

        }
    }

    private scanCallExpression(expr: CallExpression | BinaryExpression, scope: Scope) {
        for (const arg of expr.arguments) {
            this.scanInstruction(arg, scope);
        }
    }

    private scanImplDeclaration(impl: ImplDeclaration, scope: Scope) {
        this.resolver.resolveTypeEntity({
            label: impl.target,
            scope,
            resolver: type => {
                impl.targetID = type.id;
                impl.functions.forEach(fn => this.scanFn(fn, type.scope, type.id));
            }
        })
    }

    private scanPropertyAccessExpression(expr: PropertyAccessExpression, scope: Scope) {
        const walk = async (expr: PropertyAccessExpression | Identifier, scope: Scope): Promise<Scope> =>
            new Promise<Scope>(async res => {
                if (expr.kind === "identifier") {
                    return this.resolver.resolveIdentifier({
                        label: expr.label,
                        scope,
                        resolver: e => {
                            expr.id = e.id;
                            res(e.scope);
                        }
                    });
                }

                const arg2Scope = await walk(expr.arguments[0], scope);
                return walk(expr.arguments[1], arg2Scope);
            });

        walk(expr, scope);
    }

    private scanVariableDeclaration(vr: VariableDeclaration, scope: Scope) {
        if (vr.initializer) this.scanInstruction(vr.initializer, scope);

        const id = this.entities.add({
            kind: "local",
            label: vr.label,
            flags: vr.flags,
            mutable: vr.flags.includes("var"),
            scope,
            index: scope.localsCount() + 1
        });

        scope.addLocal(id);
        vr.id = id;
    }

    private scanIdentifier(ident: Identifier, scope: Scope) {
        this.resolver.resolveIdentifier({
            label: ident.label, scope,
            resolver: e => ident.id = e.id
        });
    }

    private scanType(type: TypeDeclaration | StructDeclaration | EnumDeclaration, scope: Scope) {
        const id = this.entities.add({
            kind: "type",
            label: type.label,
            flags: type.flags,
            // TODO distinguish static and instance namespaces
            scope: scope.newSubScope(),
        });
        scope.add(id);
        if (type.flags.includes("pub")) scope.export(id);
        type.id = id;
    }

    private scanFn(fn: FunctionDeclaration, outerScope: Scope, self?: string) {
        const fnScope = outerScope.newSubScope();
        fnScope.isFnScope = true;
        const id = this.entities.add({
            kind: "function",
            parameters: [],
            label: fn.label,
            flags: fn.flags,
            scope: fnScope
        });

        fn.id = id;
        outerScope.add(id);
        this.resolveFnParameters(fn, fnScope, self);

        if (fn.flags.includes("pub")) outerScope.export(id);
        if (fn.expression) this.scanInstruction(fn.expression, fnScope);
        if (fn.returnType) {
            this.resolver.resolveTypeEntity({
                scope: outerScope,
                label: fn.returnType.label,
                resolver: (e) => this.entities.update({ id, returnType: e.id })
            });
        }
    }

    private resolveFnParameters(fn: FunctionDeclaration, fnScope: Scope, self?: string) {
        const parameters: string[] = [];

        if (self) {
            const paramID = this.entities.add({
                kind: "local",
                label: "self",
                flags: [],
                scope: fnScope,
                typeEntity: self,
                index: 0,
                mutable: fn.flags.includes("mut")
            });

            fnScope.add(paramID);
            parameters.push(paramID);
        }

        fn.parameters.forEach(p => {
            const id = this.entities.add({
                kind: "local",
                flags: [],
                label: p.label,
                scope: fnScope,
                mutable: false,
                index: parameters.length
            });

            this.resolver.resolveTypeEntity({
                label: p.type!.label,
                scope: fnScope,
                resolver: entity => this.entities.update({ id, typeEntity: entity.id })
            });

            fnScope.add(id);
            parameters.push(id);
        });

        return parameters;
    }
}
