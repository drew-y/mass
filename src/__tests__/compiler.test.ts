import {
  e2eVoidText,
  gcVoidText,
  genericsText,
  goodTypeInferenceText,
  tcoText,
} from "./fixtures/e2e-file.js";
import { compile } from "../compiler.js";
import { describe, test, vi } from "vitest";
import assert from "node:assert";
import { getWasmFn, getWasmInstance } from "../lib/wasm.js";
import * as rCallUtil from "../assembler/return-call.js";

describe("E2E Compiler Pipeline", () => {
  test("Compiler can compile and run a basic void program", async (t) => {
    const mod = await compile(e2eVoidText);
    const instance = getWasmInstance(mod);
    const fn = getWasmFn("main", instance);
    assert(fn, "Function exists");
    t.expect(fn(), "Main function returns correct value").toEqual(55);
  });

  test("Compiler has good inference", async (t) => {
    const mod = await compile(goodTypeInferenceText);
    const instance = getWasmInstance(mod);
    const fn = getWasmFn("main", instance);
    assert(fn, "Function exists");
    t.expect(fn(), "Main function returns correct value").toEqual(55n);
  });

  test("Compiler can compile gc objects and map correct fns", async (t) => {
    const mod = await compile(gcVoidText);
    const instance = getWasmInstance(mod);
    t.expect(mod.validate(), "Module is valid");
    const test1 = getWasmFn("test1", instance);
    const test2 = getWasmFn("test2", instance);
    const test3 = getWasmFn("test3", instance);
    const test4 = getWasmFn("test4", instance);
    const test5 = getWasmFn("test5", instance);
    const test6 = getWasmFn("test6", instance);
    const test7 = getWasmFn("test7", instance);
    assert(test1, "Test1 exists");
    assert(test2, "Test2 exists");
    assert(test3, "Test3 exists");
    assert(test4, "Test4 exists");
    assert(test5, "Test5 exists");
    assert(test6, "Test6 exists");
    assert(test7, "Test7 exists");
    t.expect(test1(), "test 1 returns correct value").toEqual(13);
    t.expect(test2(), "test 2 returns correct value").toEqual(1);
    t.expect(test3(), "test 3 returns correct value").toEqual(2);
    t.expect(test4(), "test 4 returns correct value").toEqual(52);
    t.expect(test5(), "test 5 returns correct value").toEqual(52);
    t.expect(test6(), "test 6 returns correct value").toEqual(21);
    t.expect(test7(), "test 7 returns correct value").toEqual(-1);
  });

  test("Compiler can do tco", async (t) => {
    const spy = vi.spyOn(rCallUtil, "returnCall");
    await compile(tcoText);
    const did = spy.mock.calls.some((call) => call[1].startsWith("fib"));
    t.expect(did);
  });

  test("Generic fn compilation", async (t) => {
    const mod = await compile(genericsText);
    const instance = getWasmInstance(mod);
    const main = getWasmFn("main", instance);
    assert(main, "Main exists");
    t.expect(main(), "main 1 returns correct value").toEqual(143);
  });
});
