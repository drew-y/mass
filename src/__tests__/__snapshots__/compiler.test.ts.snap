// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`E2E Compiler Pipeline > Compiler can do tco 1`] = `
"(module
 (type $0 (func (param i32 i32 i32) (result i32)))
 (memory $0 1 150)
 (export "buffer" (memory $0))
 (export "fib" (func $fib#15065))
 (func $fib#15065 (type $0) (param $0 i32) (param $1 i32) (param $2 i32) (result i32)
  (if (result i32)
   (local.get $0)
   (then
    (return_call $fib#15065
     (i32.sub
      (local.get $0)
      (i32.const 1)
     )
     (local.get $2)
     (i32.add
      (local.get $1)
      (local.get $2)
     )
    )
   )
   (else
    (local.get $1)
   )
  )
 )
)
"
`;
